### DISCUSSION TITLE
SMTP Problems in combination with office 365

### DISCUSSION THREAD

**Discussion by ohrbeit** [dani-garcia/vaultwarden#1861] [Q&A]
### Subject of the issue

Whenever I try to send smtp mails to office365, I get the no compatible auth mechanism was found error. 
It worked like a charm before, with the same settings. I also contacted microsoft and they told me that the settings are 100% correct, we even tried to send a mail via telnet from the same container - works 100%. 

Seems that there is some problem with the smtp mailer included in the docker container. 

I've tried all possible ports and ssl / tls settings and combinations - no luck so far. 

My Problem is, I can't confirm invitations from the admin panel. I need the emails to work in order to create new working accounts. Is there any way to confirm an account from the admin panel? In that case, I could live without a working smtp mailer!

```
[2021-07-20 08:57:04.680][request][INFO] POST /api/accounts/verify-email                                                              
[2021-07-20 08:57:04.784][vaultwarden::mail][ERROR] SMTP Client error: internal client error: No compatible authentication mechanism w
as found                                                                                                                              
[2021-07-20 08:57:04.785][vaultwarden::api::core::accounts][ERROR] Error sending verify_email email: SMTP Client error: internal clien
t error: No compatible authentication mechanism was found                                                                             
[2021-07-20 08:57:04.785][response][INFO] POST /api/accounts/verify-email (post_verify_email) => 200 OK                               
^
```

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.22.1
* Web-vault version: v2.20.4b
* Running within Docker: true
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, ADMIN_TOKEN, SMTP_HOST, SMTP_PORT, SMTP_FROM

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*********.*******.**",
  "domain_origin": "*****://*********.*******.**",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": "*********.*******.**",
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": false,
  "signups_domains_whitelist": "*******.**",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "*********@*******.**",
  "smtp_from_name": "Bitwarden",
  "smtp_host": "*******-**.****.**********.*******.***",
  "smtp_password": "***",
  "smtp_port": 25,
  "smtp_ssl": false,
  "smtp_timeout": 15,
  "smtp_username": "",
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
</details>
```

**Comment by BlackDex** [comment]
Not sure which confirmations you mean, if those are the once invited via the organization, then you need to confirm those via that same route. No need for SMTP for that to work.

Regarding the SMTP, you should use these settings (i'm using office365 here):
```ini
SMTP_HOST=smtp.office365.com
SMTP_PORT=587
SMTP_SSL=true
SMTP_EXPLICIT_TLS=false
SMTP_FROM=user@office365-domain.tld
SMTP_USERNAME=user@office365-domain.tld
SMTP_PASSWORD='Very^Secret@PassW0rd#With*Some!Special^Characters&'
SMTP_AUTH_MECHANISM="Login"
```

**Comment by ohrbeit** [comment]
> Not sure which confirmations you mean, if those are the once invited via the organization, then you need to confirm those via that same route. No need for SMTP for that to work.

If I invite a new user via organisation, the user can't log in until they confirmed the email adress. In the backend, I can only re-send the invitation, but I can't confirm the mail from admin side. 

So I need smtp to work for sending out invitation mails. Or am I missing something? 

Unfortunately I've already tried those settings with a specific user account - wasn't working for me :/

**Comment by BlackDex** [comment]
> > Not sure which confirmations you mean, if those are the once invited via the organization, then you need to confirm those via that same route. No need for SMTP for that to work.
> 
> If I invite a new user via organisation, the user can't log in until they confirmed the email adress. In the backend, I can only re-send the invitation, but I can't confirm the mail from admin side.

This needs to be done via the Organization view within the web-vault, not the `/admin` interface.
See: https://bitwarden.com/help/article/managing-users/

> 
> So I need smtp to work for sending out invitation mails. Or am I missing something?
> 
> Unfortunately I've already tried those settings with a specific user account - wasn't working for me :/

Then I think SMTP Auth is disabled for your Office365 environment/user, see:
https://docs.microsoft.com/en-us/exchange/clients-and-mobile-in-exchange-online/authenticated-client-smtp-submission
https://www.limilabs.com/blog/office365-temporary-server-error-please-try-again-later-prx4

Also, be sure that you have the password quoted using single quotes if you configured that outside of the `/admin` interface.
If you configured this via the `/admin` interface, just copy/paste the password in the correct field.
And, the `SMTP_FROM` and `SMTP_USERNAME` need to be the same, else it will fail if i'm correct.

**Comment by ohrbeit** [comment]
> This needs to be done via the Organization view within the web-vault, not the `/admin` interface.
> See: https://bitwarden.com/help/article/managing-users/

I know, but as it's stated there, it is a 3 step process:
1. invitation is sent out from the organization view
2. user has to confirm / accept that invitation (here we have the problem - that invitation mail is not sent out correctly, so i can't proceed to
3. I have to confirm that user once again from the organisation view.

Problem is, that I cannot accept the invation on behalf of the user. And they can't accept it, cause they don't get the mail. 

What I would need is some sort of setting / button to skip the 3 step invitation process or to confirm the email for the user. Otherwise I would need SMTP to work.

> > So I need smtp to work for sending out invitation mails. Or am I missing something?
> > Unfortunately I've already tried those settings with a specific user account - wasn't working for me :/
> 
> Then I think SMTP Auth is disabled for your Office365 environment/user, see:
> https://docs.microsoft.com/en-us/exchange/clients-and-mobile-in-exchange-online/authenticated-client-smtp-submission
> https://www.limilabs.com/blog/office365-temporary-server-error-please-try-again-later-prx4
> 
> Also, be sure that you have the password quoted using single quotes if you configured that outside of the `/admin` interface.
> If you configured this via the `/admin` interface, just copy/paste the password in the correct field.
> And, the `SMTP_FROM` and `SMTP_USERNAME` need to be the same, else it will fail if i'm correct.

When I try it via Login, sending directly to office365 from a regular exchange user account I get:

`
[2021-07-20 14:00:42.359][error][ERROR] Smtp.
--
[CAUSE] lettre::transport::smtp::Error {
 kind: Connection,
source: "Could not connect",
`

**Comment by BlackDex** [comment]
Well, checking your output which you generated via the `diagnostics` page, it looks to me that you have it configured to port `25`, and not port `587`. And also `SMTP_SSL` is set to false.
So, check your configuration again.

**Comment by ohrbeit** [comment]
> Well, checking your output which you generated via the `diagnostics` page, it looks to me that you have it configured to port `25`, and not port `587`. And also `SMTP_SSL` is set to false.
> So, check your configuration again.

Unfortunately i've already tried both setups. 
Configuring it to send directly to the internal office 365 mail system of the company `company-en.mail.protection.outlook.com` (using port `25`, without any login)

 and also with `SMTP_SSL True` and Port `587` + a correct office 365 exchange account with full permission and login credentials to `smtp.office365.com` as host

`SMTP_HOST=smtp.office365.com
SMTP_PORT=587
SMTP_SSL=true
SMTP_EXPLICIT_TLS=false
SMTP_FROM=user@office365-domain.tld
SMTP_USERNAME=user@office365-domain.tld
SMTP_PASSWORD='Very^Secret@PassW0rd#With*Some!Special^Characters&'
SMTP_AUTH_MECHANISM="Login"`

**Comment by BlackDex** [comment]
port 25 can't be used to send email and login. That isn't allowed.

**Comment by ohrbeit** [comment]
I still can't get it to work properly... I really don't know the problem, tried all the ports, all different smtp settings - i don't know any more solutions. 

Quick Question: Is there any way (database access, admin panel, etc.) to manually accept the invitation or verify the user at all? 
So I just need to give the username and password to the co-worker as admin? 

Best Regards

**Comment by RobinDadswell** [comment]
I am also having an issue with this; 

My error message is: 
2022-05-05 21:25:08.437][vaultwarden::mail][ERROR] SMTP Connection error: error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:332:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.24.0
* Web-vault version: v2.25.1
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: false
* HTTPS Check: false
* Database type: PostgreSQL
* Database version: PostgreSQL 14.2 (Debian 14.2-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** SIGNUPS_VERIFY, SIGNUPS_DOMAINS_WHITELIST, SHOW_PASSWORD_HINT, ADMIN_TOKEN, SMTP_HOST, SMTP_SSL, SMTP_EXPLICIT_TLS, SMTP_PORT, SMTP_FROM, SMTP_USERNAME, SMTP_PASSWORD, SMTP_AUTH_MECHANISM


```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "**********://***********:***********@**.**.**.**:****/****************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "****://*********",
  "domain_origin": "****://*********",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_org_name": ".............",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "********.***,********.*****",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "Login",
  "smtp_debug": false,
  "smtp_explicit_tls": true,
  "smtp_from": "*********@********.*****",
  "smtp_from_name": "bitwarden",
  "smtp_host": "****.*********.***",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "*****@********.*****",
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Comment by BlackDex** [comment]
@RobinDadswell make sure you set `SMTP_EXPLICIT_TLS` to false.
